mcstas-comps/monitors/Monitor_nD.comp

```diff
@@ -483,23 +483,53 @@ TRACE
-        y1 = y+t1*vy;
-        if (fabs(y0) >= Vars.Cylinder_Height/2) t0 = t1;
-        if (fabs(y1) >= Vars.Cylinder_Height/2) t1 = t0;
-      }
-      if (t0 < 0 && t1 > 0)
-        t0 = t;  /* neutron was already inside ! */
-      if (t1 < 0 && t0 > 0) /* neutron exit before entering !! */
-        t1 = t;
-      /* t0 is now time of incoming intersection with the detection area */
-      if ((Vars.Flag_Shape < 0) && (t1 > 0))
-        PROP_DT(t1); /* t1 outgoing beam */
-      else
-        PROP_DT(t0); /* t0 incoming beam */
+        if (abs(Vars.Flag_Shape) == DEFS.SHAPE_BANANA) {
+            if (intersect == 1) { // Entered and left through sides
+                if (t0 < 0 && t1 > 0) {
+                    t0 = t;  /* neutron was already inside ! */
+                }
+                if (t1 < 0 && t0 > 0) { /* neutron exit before entering !! */
+                    t1 = t;
+                }
+                /* t0 is now time of incoming intersection with the detection area */
+                if ((Vars.Flag_Shape < 0) && (t1 > 0)) {
+                    PROP_DT(t1); /* t1 outgoing beam */
+                } else {
+                    PROP_DT(t0); /* t0 incoming beam */
+                }
+            } else if (intersect == 3 || intersect == 5) { // Entered from top or bottom, left through side
+                if ((Vars.Flag_Shape < 0) && (t1 > 0)) {
+                    PROP_DT(t1); /* t1 outgoing beam */
+                } else {
+                    intersect=0;
+                    Flag_Restore=1;
+                }
+            } else if (intersect == 9 || intersect == 17) { // Entered through side, left from top or bottom
+                if ((Vars.Flag_Shape < 0) && (t1 > 0)) {
+                    intersect=0;
+                    Flag_Restore=1;
+                } else {
+                    PROP_DT(t0); /* t0 incoming beam */
+                }
+            } else if (intersect == 13 || intersect == 19) { // Went through top/bottom on entry and exit
+                intersect=0;
+                Flag_Restore=1;
+            } else {
+                printf("Cylinder_intersect returned unexpected value %l\n", intersect);
+            }
+        } else {
+            // All other shapes than the BANANA 
+            if (t0 < 0 && t1 > 0)
+              t0 = t;  /* neutron was already inside ! */
+            if (t1 < 0 && t0 > 0) /* neutron exit before entering !! */
+              t1 = t;
+            /* t0 is now time of incoming intersection with the detection area */
+            if ((Vars.Flag_Shape < 0) && (t1 > 0))
+              PROP_DT(t1); /* t1 outgoing beam */
+            else
+              PROP_DT(t0); /* t0 incoming beam */
+        }
+      
       /* Final test if we are on lid / bottom of banana/sphere */
       if (abs(Vars.Flag_Shape) == DEFS.SHAPE_BANANA || abs(Vars.Flag_Shape) == DEFS.SHAPE_SPHERE) {
         if (Vars.Cylinder_Height && fabs(y) >= Vars.Cylinder_Height/2 - FLT_EPSILON) {
```

Comment:
We could remove the FLT_EPSILON on this check for BANANA now, as it is effectively just checking if the path propagated with gravity sent the neutron outside of the detector area.
I don't understand the need for lid/bottom of the sphere mentioned in the comment, so have kept this code.